lib/src/working_copy.rs

```diff
@@ -444,12 +445,30 @@ impl TreeState {
         &self,
         path: &RepoPath,
         disk_path: &Path,
-    ) -> Result<FileId, SnapshotError> {
+        max_size: Option<u64>,
+    ) -> Result<Option<FileId>, SnapshotError> {
         let file = File::open(disk_path).map_err(|err| SnapshotError::IoError {
             message: format!("Failed to open file {}", disk_path.display()),
             err,
         })?;
-        Ok(self.store.write_file(path, &mut Box::new(file))?)
+        if let Some(max_size) = max_size {
+            let meta = file.metadata().map_err(|err| SnapshotError::IoError {
+                message: format!("Failed to access metadata of file {}", disk_path.display()),
+                err,
+            })?;
+            if meta.len() > max_size {
+                tracing::warn!(
+                    "not tracking new {} byte file {}, which is larger than configured \
+                     snapshot.max-new-file-size {}",
+                    meta.len(),
+                    disk_path.display(),
+                    max_size
+                );
+                return Ok(None);
```

Comment:
Silently ignoring files is a bit scary. Maybe we can return a list of ignored paths to CLI and print them, but it might be out of scope of this PR.

Reply:
I agree. Especially since we're setting the config by default, I think we need to inform the user about ignored paths.

Reply:
That's the purpose of the `warn!` line immediately above. Are those not surfaced to the user?

Reply:
No, all `tracing` calls are just for debugging (you need `RUST_LOG=...` or `--verbose` to see it).

Reply:
It looks like we display error level tracing logs by default:

https://github.com/martinvonz/jj/blob/eac90fd113401be634538e32d9bbf6f66737aa70/src/cli_util.rs#L348

Is the thinking that we don't want critical information to be only available by a sidechannel that might e.g. easily get lost in a GUI frontend? I'd think error-level diagnostics would by definition merit visibility, though...

Reply:
I do consider this a hard blocker for this feature -- we should never silently ignore files except at express user request.

Reply:
> Is the thinking that we don't want critical information to be only available by a sidechannel that might e.g. easily get lost in a GUI frontend?

Yes, we want to provide that information in a way that any kind of UI can present.

Reply:
Right, my observation is that's probably true for any error. I guess we shouldn't be using `error!` in lib code if possible.

Reply:
Yes, except just before panicking, maybe

Reply:
Btw, it might be better to report these skipped files via a callback than to return them. That way the UI can provide feedback sooner.

Reply:
Are we targeting cases where snapshots take a long time, despite skipping work for large files?

Reply:
We already have progress reporting because snapshotting can take a while. But you're right that it's not very common, and should become less common with this PR, so do whatever is simpler, I guess.

Reply:
I've refactored this to generate an error for now, since on review I couldn't think of a reason why this specifically shouldn't be fatal, and doing so retains a simpler API.